architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,382 @@
+- We will follow user expectations for Kubernetes controllers and resources when designing the BicepDeployment controller and resource. This includes building in support for retries, status updates, and error handling.
+
+### Non-goals
+
+**Non-goal (out of scope): Full support for GitOps**
+
+- We will not yet be implementing automatic generation of BicepDeployment resources from Bicep manifests or querying Git repositories. This design will enable this work, and it will be covered in a future design document.
+
+### User scenarios
+
+#### Jon can deploy cloud (Azure or AWS) resources defined in Bicep manifests using Kubernetes tools
+
+Jon is an infrastructure operator for an enterprise company. His team manages a production application that is deployed on Kubernetes and uses cloud resources. Jon wants to declare all dependencies for his application in Kubernetes manifests, including cloud resources while leveraging Radius for its useful features. He installs Radius on his cluster and uses the rad CLI to generate a custom resource from his Bicep manifest. Jon applies the custom resource to his cluster, and Radius deploys the cloud resources defined in the Bicep manifest. If he wants to update or delete the cloud resources, he can do so by re-applying the updated custom resource or deleting the custom resource.
+
+#### Jon can deploy Radius resources defined in Bicep manifests using Kubernetes tools
+
+Now that he can see that Radius can deploy cloud resources defined in Bicep manifests, Jon wants to take advantage of Radius tooling, such as the App Graph, and fully "Rad-ify" his application. He writes a Bicep manifest that defines a Radius container that connects to the cloud resources, and uses the rad CLI to generate a custom resource from the Bicep manifest. Jon applies the custom resource to his cluster, and Radius deploys the Radius resources defined in the Bicep manifest. Now, Jon can take advantage of Radius tooling, such as the Radius Dashboard and App Graph, to manage his application.
+
+## User Experience
+
+### Radius Container
+
+**Sample Input:**
+
+#### `app.bicep`
+```bicep
+extension radius
+
+param application string
+param tag string
+param port int
+
+resource container 'Applications.Core/containers@2023-10-01-preview' = {
+  name: 'container'
+  properties: {
+    application: application
+    container: {
+      image: 'ghcr.io/radius-project/samples/demo:${tag}'
+      ports: {
+        web: {
+          containerPort: port
+        }
+      }
+    }
+  }
+}
+```
+
+#### `app.bicepparam`
+```bicep
+using 'app.bicep'
+
+param application = ''
+param tag = ''
+param port = 3000
+```
+
+**Sample Output:**
+```shell
+rad bicep generate-kubernetes app.bicep --parameters @app.bicepparam --parameters tag=latest --parameters appnetworking.bicepparam --outfile app.yaml
+
+Generating BicepDeployment resource...
+BicepDeployment resource generated at app.yaml
```

Comment:
Should we call this ARMDeployment instead of BicepDeployment? It contains ARM compilation from Bicep, but it is not Bicep.

Reply:
Please no references to a cloud provider-specific components in a cross-platform tool.

Reply:
Ideas:

ARMDeployment
TemplateDeployment
RadiusDeployment
BicepDeployment

Reply:
DeploymentTemplate